File: python/searching/binary-search.py
import math

def binary_search(arr, l, r, x):
  if (r >= l):
    mid = l + math.floor((r-l)/2)

    if (arr[mid] == x):
      return mid
    
    if (arr[mid] > x):
      return binary_search(arr, l, mid - 1, x)
    
    if (arr[mid] < x):
      return binary_search(arr, mid + 1, r, x)

  return None

arr = [ 2, 3, 4, 50, 40 ]
arr.sort()
x = 2
result = binary_search(arr, 0, len(arr), x)
print('Search Element is found', result != None)

File: javascript/searching/binary-search.js

function binarySearch(arr, l, r, searchElement) {
  
  if (r >= l) {
    let mid = l + Math.floor((r - l) / 2); //?

    if (arr[mid] === searchElement) {
      return mid;
    }

    if (arr[mid] > searchElement) {
      return binarySearch(arr, l, mid - 1, searchElement);
    }

    if (arr[mid] < searchElement) {
      return binarySearch(arr, mid+1, r, searchElement);
    }
  }

  return -1;
}

const arr = [ 2, 3, 4, 10, 40 ];
const searchElement = 10;

const sortedArr = arr.sort((a,b) => a < b);
const totalLength = sortedArr.length; //?
binarySearch(arr,0, totalLength-1, searchElement); //?